Panic on Tx rollback when connection has been closed by PgBouncer/Postgres (close of closed channel)

**Describe the bug**

`Tx.Rollback()` consistently panics when the underlying connection has been closed by the server (PgBouncer or Postgres) due to an idle-in-transaction timeout.

Stack trace:

```
panic: close of closed channel

goroutine 3122 [running]:
github.com/jackc/pgx/v5/pgconn.(*PgConn).receiveMessage(0xc000419b08)
        /home/gitlab-runner/go/pkg/mod/github.com/jackc/pgx/v5@v5.7.0/pgconn/pgconn.go:583 +0x370
github.com/jackc/pgx/v5/pgconn.(*Pipeline).getResults(0xc000419c60)
        /home/gitlab-runner/go/pkg/mod/github.com/jackc/pgx/v5@v5.7.0/pgconn/pgconn.go:2162 +0x34
github.com/jackc/pgx/v5/pgconn.(*Pipeline).Close(0xc000419c60)
        /home/gitlab-runner/go/pkg/mod/github.com/jackc/pgx/v5@v5.7.0/pgconn/pgconn.go:2263 +0xf6
github.com/jackc/pgx/v5.(*Conn).deallocateInvalidatedCachedStatements(0xc000dceb40, {0xdde860, 0xc0006c0e00})
        /home/gitlab-runner/go/pkg/mod/github.com/jackc/pgx/v5@v5.7.0/conn.go:1402 +0x226
github.com/jackc/pgx/v5.(*Conn).Exec(0xc000dceb40, {0xdde860?, 0xc0006c0e00?}, {0xcc6678, 0x8}, {0x0, 0x0, 0x0})
        /home/gitlab-runner/go/pkg/mod/github.com/jackc/pgx/v5@v5.7.0/conn.go:462 +0xcc
github.com/jackc/pgx/v5.(*dbTx).Rollback(0xc000012c00, {0xdde860?, 0xc0006c0e00?})
        /home/gitlab-runner/go/pkg/mod/github.com/jackc/pgx/v5@v5.7.0/tx.go:204 +0x4b
github.com/jackc/pgx/v5/pgxpool.(*Tx).Rollback(0xc000012c18, {0xdde860?, 0xc0006c0e00?})
        /home/gitlab-runner/go/pkg/mod/github.com/jackc/pgx/v5@v5.7.0/pgxpool/tx.go:38 +0x2a
```

I suspect this a repeat of the issue identified back in https://github.com/jackc/pgx/issues/1920

**To Reproduce**
Steps to reproduce the behavior:

1. Initialize a `PgxPool` and connect to a PostgreSQL instance behind a PgBouncer proxy.
2. Initiate a transaction: `tx, _ := pool.Begin(ctx)`.
3. Hold the transaction for the exact same amount of time the PgBouncer proxy and PostgreSQL database are configured to terminate idle-in-transaction connections.
4. Attempt to commit something to the database using an expired context. This will result, as expected, in a "context deadline exceeded" error:
```go
err := tx.Commit(ctx) // err is context.DeadlineExceeded
```
5. Now, attempt a rollback of the transaction using a different context. For instance, using the background context as is generally considered best practice. This causes a panic.
```go
tx.Rollback(context.Background()) // panic!!
```

Unfortunately, I have been unable to reproduce this in a unit or integration test, given the toil involved in configuring PgBouncer for testing. Following the above steps with only PostgreSQL (no PgBouncer) does not result in the panic, which leads me to believe PgBouncer is a key factor here.

I have been able to reproduce this reliably in a staging environment, inside a Kubernetes pod and connecting to a proper PgBouncer + PostgreSQL setup. I can't really reproduce the full code here due to the amount of boilerplate necessary to run this in said K8s pod, but the general structure of the relevant code segment is:

```go
// Both PgBouncer and PostgreSQL are configured to terminate idle transactions at
// the 1 minute mark.
cacheCtx, cacheCtxCancel := context.WithTimeout(context.Background(), time.Minute)
defer cacheCtxCancel()
tx, err := pool.Begin(cacheCtx)
if err != nil {
    panic(err.Error())
}

time.Sleep(time.Minute)
// at this point both the cache context and the idle-in-transaction deadlines
// should have been hit.

_, err = tx.Exec(cacheCtx, "INSERT INTO random_table (value) VALUES ($1);", 123)
if err != context.DeadlineExceeded {
    panic(fmt.Sprintf("%v", err))
}
err = tx.Rollback(context.Background()) // panics!!
```

**Expected behavior**
Rollback after a connection has been closed by the server should return an error, but never panic.

**Actual behavior**
Rollback after a connection has been closed by the server results in a panic in `(*PgConn).receiveMessage`: `close of closed channel`.

**Version**
 - Go: `go version go1.25.5 linux/amd64`.
 - PostgreSQL: PostgreSQL 14.8
 - pgx: v5.7.0

**Additional context**

I suspect this a regression of behavior identified and fixed back in https://github.com/jackc/pgx/issues/1920

Looking at the stack traces and the code, I note that the last three calls in the stack traces of these two issues are almost exactly the same, except in my case `(*Pipeline).Close` calls `(*Pipeline).getResults`, whereas in https://github.com/jackc/pgx/issues/1920 (pgx version v5.4.3), `(*Pipeline).Close` used to instead call `(*Pipeline).GetResults` (note the capitalization!!). `(*Pipeline).GetResults` [contains a check for the `p.closed` flag](https://github.com/jackc/pgx/blob/v5.7.0/pgconn/pgconn.go#L2146) used in the fix for https://github.com/jackc/pgx/issues/1920 (https://github.com/jackc/pgx/commit/2e84dccaf57b4fa803149884f2149dfa9e923593), whereas `(*Pipeline).getResults` _does not include this check_.

I believe this regression may have accidentally been introduced in https://github.com/jackc/pgx/commit/22fe50149b6f723af4d77c71244d54964bd14e03#diff-7600c6d53b2bb66dc26f15ff10fb080ad09158eea6e7562a8dc1db33d1ea2abfL2181.

--- a/pgconn/pgconn.go
+++ b/pgconn/pgconn.go
@@ -584,6 +584,10 @@ func (pgConn *PgConn) peekMessage() (pgproto3.BackendMessage, error) {
 
 // receiveMessage receives a message without setting up context cancellation
 func (pgConn *PgConn) receiveMessage() (pgproto3.BackendMessage, error) {
+	if pgConn.status == connStatusClosed {
+		return nil, &connLockError{status: "conn closed"}
+	}
+
 	msg, err := pgConn.peekMessage()
 	if err != nil {
 		return nil, err
@@ -2821,7 +2825,7 @@ func (p *Pipeline) Close() error {
 	}
 
 	for p.state.ExpectedReadyForQuery() > 0 {
-		_, err := p.getResults()
+		results, err := p.getResults()
 		if err != nil {
 			p.err = err
 			var pgErr *PgError
@@ -2829,6 +2833,15 @@ func (p *Pipeline) Close() error {
 				p.conn.asyncClose()
 				break
 			}
+		} else if results == nil {
+			// getResults returns (nil, nil) when the request queue is exhausted but
+			// ExpectedReadyForQuery is still > 0. This can happen when FATAL errors consume
+			// queued request slots without the server ever sending ReadyForQuery.
+			p.conn.asyncClose()
+			if p.err == nil {
+				p.err = errors.New("pipeline: no more results but expected ReadyForQuery")
+			}
+			break
 		}
 	}
 

--- a/pgconn/pgconn_test.go
+++ b/pgconn/pgconn_test.go
@@ -4389,6 +4389,110 @@ func TestFatalErrorReceivedInPipelineMode(t *testing.T) {
 	require.Error(t, err)
 }
 
+// https://github.com/jackc/pgx/issues/2470
+// When the server sends multiple FATAL errors in a single batch (as PgBouncer can do when
+// terminating idle-in-transaction connections), Pipeline.Close() must not panic with
+// "close of closed channel" on cleanupDone. The first FATAL triggers OnPgError which closes
+// the connection and cleanupDone. The second FATAL, still in the read buffer, must not
+// attempt to close cleanupDone again.
+//
+// This test sends all server responses in a single TCP write to guarantee both FATAL errors
+// are in the chunkReader buffer simultaneously.
+func TestPipelineCloseDoesNotPanicOnMultipleFatalErrors(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
+	defer cancel()
+
+	ln, err := net.Listen("tcp", "127.0.0.1:")
+	require.NoError(t, err)
+	defer ln.Close()
+
+	serverErrChan := make(chan error, 1)
+	go func() {
+		defer close(serverErrChan)
+
+		conn, err := ln.Accept()
+		if err != nil {
+			serverErrChan <- err
+			return
+		}
+		defer conn.Close()
+
+		err = conn.SetDeadline(time.Now().Add(59 * time.Second))
+		if err != nil {
+			serverErrChan <- err
+			return
+		}
+
+		backend := pgproto3.NewBackend(conn, conn)
+
+		// Handle startup
+		_, err = backend.ReceiveStartupMessage()
+		if err != nil {
+			serverErrChan <- err
+			return
+		}
+		backend.Send(&pgproto3.AuthenticationOk{})
+		backend.Send(&pgproto3.BackendKeyData{ProcessID: 0, SecretKey: 0})
+		backend.Send(&pgproto3.ReadyForQuery{TxStatus: 'I'})
+		err = backend.Flush()
+		if err != nil {
+			serverErrChan <- err
+			return
+		}
+
+		// Read all client pipeline messages (Parse, Describe, Parse, Describe, Sync)
+		for i := 0; i < 5; i++ {
+			_, err = backend.Receive()
+			if err != nil {
+				serverErrChan <- err
+				return
+			}
+		}
+
+		// Send ALL responses in a single write so they all end up in the chunkReader buffer.
+		// This simulates PgBouncer sending a FATAL and then the real PostgreSQL also sending
+		// a FATAL, both arriving in the same TCP segment.
+		backend.Send(&pgproto3.ParseComplete{})
+		backend.Send(&pgproto3.ParameterDescription{})
+		backend.Send(&pgproto3.RowDescription{Fields: []pgproto3.FieldDescription{
+			{Name: []byte("mock")},
+		}})
+		// Two FATAL errors back-to-back in the same write buffer
+		backend.Send(&pgproto3.ErrorResponse{Severity: "FATAL", Code: "57P01", Message: "terminating connection due to administrator command"})
+		backend.Send(&pgproto3.ErrorResponse{Severity: "FATAL", Code: "57P01", Message: "terminating connection due to administrator command"})
+		err = backend.Flush()
+		if err != nil {
+			serverErrChan <- err
+			return
+		}
+	}()
+
+	parts := strings.Split(ln.Addr().String(), ":")
+	host := parts[0]
+	port := parts[1]
+	connStr := fmt.Sprintf("sslmode=disable host=%s port=%s", host, port)
+
+	ctx, cancel = context.WithTimeout(ctx, 59*time.Second)
+	defer cancel()
+	conn, err := pgconn.Connect(ctx, connStr)
+	require.NoError(t, err)
+
+	pipeline := conn.StartPipeline(ctx)
+	pipeline.SendPrepare("s1", "select 1", nil)
+	pipeline.SendPrepare("s2", "select 2", nil)
+	err = pipeline.Sync()
+	require.NoError(t, err)
+
+	// Do NOT call GetResults. Call Close() directly so it drains results via getResults().
+	// The first FATAL closes the connection via OnPgError, including close(cleanupDone).
+	// The second FATAL is still buffered in chunkReader. Without the fix, processing it
+	// would attempt to close cleanupDone again, causing a panic.
+	err = pipeline.Close()
+	require.Error(t, err)
+}
+
 func mustEncode(buf []byte, err error) []byte {
 	if err != nil {
 		panic(err)

